### PR TITLE
Introduced `EventsDriven`

### PR DESCRIPTION
**Pull request by KickItLikeShika**
As explained here: https://github.com/pytorch/ignite/wiki/Roadmap#refactor-engine
This PR will separate the events registration code from the other code.

Check list:

- [x] New tests are added (if a new feature is added)
- [x] New doc strings: description and/or example code are in RST format
- [x] Documentation is updated (if required)

Code adapted from https://github.com/vfdev-5/ignite/tree/eventdriven

### CODE REVIEW

**Comment by KickItLikeShika** [comment]
Please review this as soon as you have time, @sdesrozis @trsvchn @vfdev-5. Thanks.

File: ignite/base/mixins.py
Comment (sdesrozis) [inline ignite/base/mixins.py:33]: Even if it’s a mixin, it’s an important library’s feature, as Metric for instance. It would deserve a specific file. What do you think?

File: ignite/base/mixins.py
Comment (sdesrozis) [inline ignite/base/mixins.py:45]: Could we use an ignite logger ?

File: ignite/base/mixins.py
Comment (sdesrozis) [inline ignite/base/mixins.py:202]:
```suggestion
        arguments updates arguments passed using :meth:`~ignite.base.mixins.EventsDriven.add_event_handler`.
```
As L185 ?

File: ignite/engine/engine.py
Comment (sdesrozis) [inline ignite/engine/engine.py:325]: Useless ?

File: ignite/engine/engine.py
Comment (sdesrozis) [inline ignite/engine/engine.py:281]: Same here, useless ?

File: ignite/engine/engine.py
Comment (sdesrozis) [inline ignite/engine/engine.py:317]: Same here.

File: ignite/engine/engine.py
Comment (sdesrozis) [inline ignite/engine/engine.py:340]: And finally here.

**Comment by sdesrozis** [comment]
@KickItLikeShika Thanks for this PR! It's a tricky part, and good for you for being comfortable enough with it!

I left a few comments and some tests must be fixed.

File: ignite/engine/engine.py
Comment (KickItLikeShika) [reply ignite/engine/engine.py:281]: Yes, all the functions left in the Engine is left for the keeping the docs.

File: ignite/base/mixins.py
Comment (KickItLikeShika) [reply ignite/base/mixins.py:202]: Yes that's right, thanks!

File: ignite/base/mixins.py
Comment (KickItLikeShika) [reply ignite/base/mixins.py:45]: It's the same logger used in the Engine now, do we want to change it? https://github.com/pytorch/ignite/blob/master/ignite/engine/engine.py#L126

File: ignite/base/mixins.py
Comment (KickItLikeShika) [reply ignite/base/mixins.py:33]: I agree with that, thanks!

**Comment by KickItLikeShika** [comment]
@sdesrozis can you please point out to the tests that should be fixed? only the there is one failing test that should be fixed, usually it was working fine without any updates, see here #2106, so I'm a bit surprised with this failure, do you thing there is other tests that should be updated?

**Comment by vfdev-5** [comment]
@KickItLikeShika there are many failing tests (13) which should be fixed before merging.

**Comment by KickItLikeShika** [comment]
@vfdev-5 Yes I mean just 2 unit tests should be fixed

File: ignite/base/mixins.py
Comment (sdesrozis) [reply ignite/base/mixins.py]: That's the point I would like to figure out.

File: .github/workflows/tpu-tests.yml
Comment (vfdev-5) [inline .github/workflows/tpu-tests.yml:67]: Please revert this change. I'll update `refactor-engine` branch separately.

File: ignite/engine/engine.py
Comment (sdesrozis) [inline ignite/engine/engine.py:272]: @KickItLikeShika you resolved the comment, yet you didn’t answer.

**Comment by sdesrozis** [comment]
@KickItLikeShika error in the docstring of EventDriver.

File: ignite/base/events_driven.py
Comment (vfdev-5) [inline ignite/base/events_driven.py:14]: If we make this class public and expose it in the docs. We need to make this docstring better and provide examples

**Comment by vfdev-5** [comment]
@KickItLikeShika I thought we agreed on the first PR to do only:

> - introducing EventsDriven and other mixins classes without changing anything in Engine

Here you are already modifying `Engine`.

File: tests/ignite/base/test_events_driven.py
Comment (vfdev-5) [inline tests/ignite/base/test_events_driven.py:29]: Which updates you are talking about here? I think this test file should contain all tests. If it is not the case, maybe something is wrong

File: ignite/engine/engine.py
Comment (KickItLikeShika) [reply ignite/engine/engine.py:272]: Very sorry about that, but I answered your question here https://github.com/pytorch/ignite/pull/2107#discussion_r669973702, I thought you saw it and agreed so I resolved the comment, again sorry.
I think keeping the full docs in the Engine docs and minimized docs in the base class if something nice, what do you think?

File: ignite/base/events_driven.py
Comment (KickItLikeShika) [reply ignite/base/events_driven.py:14]: I agree with that, thanks!

File: tests/ignite/base/test_events_driven.py
Comment (KickItLikeShika) [reply tests/ignite/base/test_events_driven.py:29]: Sorry I should have made that more clear in the comments, currently we are fetching the event by the same old way in the (using state) `_handler_wrapper` https://github.com/pytorch/ignite/blob/44a64d6d8def637ae636471faf1487a18d209456/ignite/base/events_driven.py#L49, which is will make this test fail. Why don't we update it and fetch the event using `_allowed_event_counts` as we should do in the upcoming updates? using `_allowed_event_counts` without the other necessary updates along with it will make `test_deterministic` fails, so I commented it for now, and as soon as we update the engine and `_handler_wrapper` I will uncomment this.

**Comment by KickItLikeShika** [comment]
@vfdev-5 can you please point out to which modification in the engine you mean? I haven't changed anything in the engine except the necessary for introducing EventsDriven

File: tests/ignite/base/test_events_driven.py
Comment (vfdev-5) [reply tests/ignite/base/test_events_driven.py:29]: > Why don't we update it and fetch the event using _allowed_event_counts as we should do in the upcoming updates?

Yes, this is also my question :) Ahmed, it is your PR


I commented before to remove EventsDriven from Engine and in this PR we should just introduce EventsDriven...
All tests should be present and passing.

File: ignite/engine/engine.py
Comment (vfdev-5) [inline ignite/engine/engine.py:20]: @KickItLikeShika Engine should not inherit from `EventsDriven` in this PR

File: tests/ignite/base/test_events_driven.py
Comment (KickItLikeShika) [reply tests/ignite/base/test_events_driven.py:29]: > using _allowed_event_counts without the other necessary updates along with it will make test_deterministic fails, so I commented it for now, and as soon as we update the engine and _handler_wrapper I will uncomment this.

That's the answer for the question. I understand your point now, I will revert the engine updates.

File: ignite/engine/events.py
Comment (vfdev-5) [inline ignite/engine/events.py:351]: @KickItLikeShika let's revert changes with the State as well. Thanks

Comment (KickItLikeShika) [reply ignite/engine/events.py:351]: What about keeping it like this for now? My point is in this PR we don't do huge changes that requires a lot of reviews from your side, so I think moving State in this PR will prevent you from more headaches in next PR, and the next PR will be more focused on the important changes, what do you think?

Comment (vfdev-5) [reply ignite/engine/events.py:351]: Let's do things by logical steps

Comment (KickItLikeShika) [reply ignite/engine/events.py:351]: I see your point, I will revert state updates.

File: ignite/base/events_driven.py
Comment (vfdev-5) [inline ignite/base/events_driven.py:74]: @KickItLikeShika please add tests to cover uncovered cases.

File: tests/ignite/base/test_mixins.py
Comment (vfdev-5) [inline tests/ignite/base/test_mixins.py:3]: I think this can be reverted.

**Review by vfdev-5 — COMMENTED** [review]
@KickItLikeShika Thanks for the updates. Remains few things and we can merge it

**Review by vfdev-5 — APPROVED** [review]
Thanks @KickItLikeShika !

File: ignite/base/events_driven.py
Comment (vfdev-5) [inline ignite/base/events_driven.py:14]:
```suggestion
    responsible for registering events and triggering events, it also keeps track of
    the allowed events, and how many times they have been triggered.
```
typos

File: ignite/base/events_driven.py
Comment (vfdev-5) [inline ignite/base/events_driven.py:17]:
```suggestion
        last_event_name: last event name triggered.
```

File: ignite/base/events_driven.py
Comment (vfdev-5) [inline ignite/base/events_driven.py:19]: This does not look good:
- https://deploy-preview-2107--pytorch-ignite-preview.netlify.app/generated/ignite.base.events_driven.eventsdriven#ignite.base.events_driven.EventsDriven

Comment (KickItLikeShika) [reply ignite/base/events_driven.py:19]: I agree that doesn't look very neat, but that's the same way we render examples in the entire codebase, should I find a way to fix it here? or leave it?

Comment (vfdev-5) [reply ignite/base/events_driven.py:19]: We may do that later everywhere. Could you please open an issue about that. Thanks !

**Pull request merged**
